client/ayon_core/hosts/houdini/api/hda_utils.py

```diff
@@ -0,0 +1,414 @@
+            # fails to resolve @sourcename var with backslashed
+            # paths correctly. So we force forward slashes
+            path = path.replace("\\", "/")
+            with _unlocked_parm(file_parm):
+                file_parm.set(path)
+
+            if node.evalParm("show_thumbnail"):
+                # Update thumbnail
+                # TODO: Cache thumbnail path as well
+                version_id = repre_entity["versionId"]
+                thumbnail_dir = node.evalParm("thumbnail_cache_dir")
+                thumbnail_path = _get_thumbnail(project_name, version_id,
+                                                thumbnail_dir)
+                set_node_thumbnail(node, thumbnail_path)
+    else:
+        # Clear filepath and thumbnail
+        with _unlocked_parm(file_parm):
+            file_parm.set("")
+        set_node_thumbnail(node, None)
+
+
+def set_node_thumbnail(node, thumbnail):
+    """Update node thumbnail to thumbnail"""
+    if thumbnail is None:
+        lib.set_node_thumbnail(node, None)
+
+    rect = compute_thumbnail_rect(node)
+    lib.set_node_thumbnail(node, thumbnail, rect)
+
+
+def compute_thumbnail_rect(node):
+    """Compute thumbnail bounding rect based on thumbnail parms"""
+    offset_x = node.evalParm("thumbnail_offsetx")
+    offset_y = node.evalParm("thumbnail_offsety")
+    width = node.evalParm("thumbnail_size")
+    # todo: compute height from aspect of actual image file.
+    aspect = 0.5625  # for now assume 16:9
+    height = width * aspect
+
+    center = 0.5
+    half_width = (width * .5)
+
+    return hou.BoundingRect(
+        offset_x + center - half_width,
+        offset_y,
+        offset_x + center + half_width,
+        offset_y + height
+    )
+
+
+def on_thumbnail_show_changed(node):
+    """Callback on thumbnail show parm changed"""
+    if node.evalParm("show_thumbnail"):
+        # For now, update all
+        on_representation_id_changed(node)
+    else:
+        lib.remove_all_thumbnails(node)
+
+
+def on_thumbnail_size_changed(node):
+    """Callback on thumbnail offset or size parms changed"""
+    thumbnail = lib.get_node_thumbnail(node)
+    if thumbnail:
+        rect = compute_thumbnail_rect(node)
+        thumbnail.setRect(rect)
+        lib.set_node_thumbnail(node, thumbnail)
+
+
+def on_representation_id_changed(node):
+    """Callback on representation id changed
+
+    Args:
+        node (hou.Node): Node to update.
+    """
+    repre_id = node.evalParm("representation")
+    set_representation(node, repre_id)
+
+
+def on_representation_parms_changed(node):
+    """
+    Usually used as callback to the project, folder, product, version and
+    representation parms which on change - would result in a different
+    representation id to be resolved.
+
+    Args:
+        node (hou.Node): Node to update.
+    """
+    project_name = node.evalParm("project_name") or get_current_project_name()
+    representation_id = get_representation_id(
```

Comment:
No - and this is on purpose actually. We want a product name change on the node to dynamically reroute the representation id - so that the artists can set up Houdini expression/variables or alike to potentially even dynamically remap say the folder path to a different context in a multi-shot workflow in Houdini.

They could say have Houdini context option set to "folderPath = 010" and switch it to 020 in a single file to have the full node tree use the context variable and get re-evaluated to the new asset or alike. The **biggest** issue yet though, is that this would only really be 'logical' with unversioned (hero versions) of products and I'm not sure yet how to really deal with multi-shot workflows in a single graph in Houdini whilst still easily maintaining version decisions per shot, etc.

Reply:
Ok new question, can this be done in bulk instead of node by node?

Reply:
It's a callback that triggers if a parm (=attribute) changes on one particular node in Houdini. That happens per node per parm so there isn't much we can do to batch these unfortunately.

Since it also only ever triggers if you tweak a specific node then performance is totally fine. However, if you were to somehow select 100s of nodes and tweak the parms all at once, then yes - you'd likely get hit by the bottleneck that it's actually doing one by one.

I guess technically we could wait in a thread until all parms are set, then go over the scene and collect any 'dirty' nodes and batch 'em in one go. But there's a big complexity overhead there - it'll be much slower for the most used case (changing a single parm on a single node) since we'd then always need to go over the full scene to find all nodes (unless we store it in a houdini session variable on change and check just that).

I understand the question - unfortunately I don't see a trivial way to resolve this particular case. Also, I don't think it'll become a production issue on short notice.

Reply:
in my prior studio we solved this by having a singleton cascading cache system that would only invalidate the downstream fields from the field that changed so upstream entities wouldn't need to be re-queried. Having a cache system is pretty crucial when dealing with Houdini callbacks as they can get triggered quite a lot and you want to be smart on lowering the amount of queries needed. Even with such system we had a few occasions where we would sometimes bring the server to a haul because of the amount of queries hitting it (of course AYON isn't being used at that scale of users but still worth considering) so we had to do some other things to break expressions on parms when submitting scenes to the farm to avoid extra queries doing rendering/caching.

Reply:
That could be doable - sure. Could be a separate PR however to keep this streamlined? I do like that - even though it does come with its own potential 'side effects' of course.

Reply:
yeah up to you, just try to be as minimal as possible on repeating the `get_*` calls. I see a few of them in @MustafaJafar code snippet he shared so just try be smart on not repeating those unnecessarily